### PR TITLE
[Service Bus Client] Add Parser, Hide Shared Key Credential

### PR DESCRIPTION
**Pull request by jsquire**
# Summary

The focus of these changes is to add the connection string parser to the public API surface, while hiding the shared key credential.  The credential types will be retired when the concept is added to Azure.Core, and the client constructors are converted over to the shared types.

**Note:** These changes have been reviewed and approved by the architecture board and the .NET language architect. Marking with the corresponding tag for tracking purposes.

# Last Upstream Rebase

Wednesday, October 29, 1:19pm (EDT)

# References
- [Implement a Service Bus Connection String Parser (#16051)](https://github.com/Azure/azure-sdk-for-net/issues/16051)
- [Accept the Azure.Core Shared Key and SAS Credentials (#16419)](https://github.com/Azure/azure-sdk-for-net/issues/16419)
- [Follow-Up: Client Creation Discussion](https://gist.github.com/jsquire/13742f53bf642ec67baf9491fc844cff)
- [Client Creation Discussion](https://gist.github.com/jsquire/55a1a934246fd6d9fe9c9e21443d7023)

### CODE REVIEW

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Resources.resx [hunk 1]
```diff
@@ -300,4 +300,7 @@
   <data name="OnlyOneSharedAccessAuthorizationMayBeSpecified" xml:space="preserve">
     <value>The authorization for a connection string may specifiy a shared key or precomputed shared access signature, but not both.  Please verify that your connection string does not have the `SharedAccessSignature` token if you are passing the `SharedKeyName` and `SharedKey`.</value>
   </data>
+  <data name="InvalidEndpointAddress" xml:space="preserve">
+    <value>The endpoint address could not be parsed; it was either malformed or not using the `sb://` scheme.</value>
```
Comment (JoshLove-msft) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Resources.resx:304]: I thought not using the sb scheme was fine because we will add it in for you:
```
                       var endpointBuilder = new UriBuilder(value)
                        {
                            Scheme = ServiceBusEndpointScheme,
                            Port = -1
                        };
```

File: sdk/servicebus/Azure.Messaging.ServiceBus/tests/SessionRecords/ServiceBusManagementClientLiveTests/SqlFilterParamsAsync.json [hunk 2]
```diff
@@ -1,18 +1,18 @@
 {
   "Entries": [
     {
-      "RequestUri": "https://recordedtests.servicebus.windows.net/d23cafe1?api-version=2017-04",
```
Comment (JoshLove-msft) [inline sdk/servicebus/Azure.Messaging.ServiceBus/tests/SessionRecords/ServiceBusManagementClientLiveTests/SqlFilterParamsAsync.json:4]: Why did we need to re-record all the tests?

**Review by JoshLove-msft — APPROVED** [review]

**Review by MiYanni — APPROVED** [review]

File: sdk/servicebus/Azure.Messaging.ServiceBus/tests/SessionRecords/ServiceBusManagementClientLiveTests/SqlFilterParamsAsync.json (hunk 2 above)
Comment (jsquire) [reply sdk/servicebus/Azure.Messaging.ServiceBus/tests/SessionRecords/ServiceBusManagementClientLiveTests/SqlFilterParamsAsync.json:4]: May not have been necessary;  I usually do a refresh when I make public API changes to the associated class.   In this case, the change was to the constructor, so it should have been non-impactful, but it's also a good forcing function to grab a new capture.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Resources.resx (hunk 1 above)
Comment (jsquire) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Resources.resx:304]: The source can use any scheme so long as the builder can successfully replace it.  It is possible to mangle the value in a way that the scheme doesn't update.  This message corresponds to a post-builder check to be sure that the scheme formatting was successful.  It's some weird corner case that is covered by one of the parser test case sets.  _(I'm being lazy and not digging into the details, but we can discuss offline and find it, if you're interested.)_

File: sdk/servicebus/Azure.Messaging.ServiceBus/tests/SessionRecords/ServiceBusManagementClientLiveTests/SqlFilterParamsAsync.json (hunk 2 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/tests/SessionRecords/ServiceBusManagementClientLiveTests/SqlFilterParamsAsync.json:4]: In general, it is better to avoid doing a re-rerecording unless the tests fail for an expected reason. Otherwise we can miss unexpected changes to the behavior.
